### ISSUE TITLE
Add torch.distributed backend for XLA to enable torch.distributed.* functions

### ISSUE THREAD

**Issue by myleott** [pytorch/xla#2091]
## 🚀 Feature
I think it'd be great to add a backend for XLA to PyTorch distributed so that the `torch.distributed.*` functions work.

## Motivation
Writing code that works both on GPU and TPU requires writing helper functions like:
```python
def all_reduce(tensor):
  global _USE_XLA
  if _USE_XLA:
    import torch_xla.core.xla_model as xm
    return xm.all_reduce('sum', [tensor])
  else:
    return torch.distributed.all_reduce(tensor)
```

It would be much more convenient if we could instead just do:
```python
torch.distributed.init_process_group(backend='xla')
torch.distributed.all_reduce(tensor)
```

## Pitch
One concern is that folks may not want XLA logic creeping into core. There is a `torch.distributed.distributed_c10d.Backend.register_backend` function, so perhaps pytorch/xla could register a new backend rather than modify PyTorch core.

**Comment by dlibenzi** [comment]
Maybe not using a global flag, as the tensor type already has the necessary info to trigger proper behavior 😉 

While a `compat` layer would not be totally unreasonable, I cannot imagine many functions in there ATM, to justify its creation.
Anything else besides `all_reduce()`?

**Comment by myleott** [comment]
> Maybe not using a global flag, as the tensor type already has the necessary info to trigger proper behavior 😉

Ah, sure, I wasn’t suggesting you add wrappers like this. Rather, it’d be better if we could use the existing torch.distributed API and have it internally call the appropriate XLA functions. My hope is that this could be done via an XLA “backend” in torch.distributed.

The functions I’m thinking are: get_rank, get_world_size, new_group (process groups in XLA are just python lists, so this is easy). Also maybe also barrier().

Another advantage of this is that code would need fewer changes to be ported to XLA, since any calls to `torch.distributed.get_world_size`, etc. wouldn’t need to be rewritten to `xm.xrt_world_size()`

**Comment by dlibenzi** [comment]
My fear doing that is the pytorch grew some APIs/behavior which do not apply to XLA, so a user suddenly expects all the torch.distributed API surface to work.

But I am OK if there is a clean hook API (means no monkey-patching 😄 ), to try that route.

**Comment by vfdev-5** [comment]
I think this can be nice feature 👍  
We do a similar things [here](https://github.com/pytorch/ignite/blob/idist/ignite/distributed/utils.py) trying to unify torch native dist and xla on basic methods like:  get_rank, get_world_size, barrier, spawn, all_reduce, all_gather... Seems to work OK.
Local rank is the only thing that is treated a bit differently between torch.distributed and xla...

**Comment by froody** [comment]
cc @pritamdamania  @mrshenli

**Comment by mrshenli** [comment]
@froody thanks for the tag. We recently started exploring this, and will post here when we have concrete updates.
